File: homework/quang_tu/lesson_5/path-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        return dfs(root, 0, targetSum, {});
    }

    vector<vector<int>> dfs(TreeNode* root, int sum, int target, vector<int> path) {
        if (!root)
            return {};

        sum += root->val;
        path.push_back(root->val);

        vector<vector<int>> result;

        if (!root->left && !root->right) {
            if (sum == target)
                result.push_back(path);
        }

        vector<vector<int>> left = dfs(root->left, sum, target, path);
        vector<vector<int>> right = dfs(root->right, sum, target, path);

        result.insert(result.end(), left.begin(), left.end());
        result.insert(result.end(), right.begin(), right.end());

        return result;
    }
};

File: homework/Long_Nguyen/lesson_7/PathSumII.java
import java.util.ArrayList;
import java.util.List;

// https://leetcode.com/problems/path-sum-ii
public class PathSumII {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null)
            return result;
        helper(root, targetSum, new ArrayList<>(), result);
        return result;
    }

    private void helper(TreeNode root, int target, List<Integer> current, List<List<Integer>> result) {
        if (root == null)
            return;

        current.add(root.val);
        if (root.left == null && root.right == null && root.val == target) {
            result.add(new ArrayList<>(current));
            current.remove(current.size() - 1);
            return;
        } else {
            helper(root.left, target - root.val, current, result);
            helper(root.right, target - root.val, current, result);
        }
        current.remove(current.size() - 1);
    }

//    public static void main(String[] args) {
//        PathSumII sol = new PathSumII();
//        TreeNode root = new TreeNode(5);
//        root.left = new TreeNode(4);
//        root.left.left = new TreeNode(11);
//        root.left.left.left = new TreeNode(7);
//        root.left.left.right = new TreeNode(2);
//
//        root.right = new TreeNode(8);
//        root.right.left = new TreeNode(13);
//        root.right.right = new TreeNode(4);
//        root.right.right.right = new TreeNode(1);
//
//        sol.pathSum(root, 22);
//    }
}
